nav2_loopback_sim/src/loopback_simulator.cpp

```diff
@@ -0,0 +1,164 @@
+  if (transform_initpose_once_ == true) {
+    try {
+      geometry_msgs::msg::TransformStamped transform = tf_buffer_->lookupTransform(
+        "odom",
+        init_pose_.header.frame_id,
+        tf2::TimePointZero);
+
+      tf2::doTransform(init_pose_, odom_updated_pose_, transform);
+    } catch (const tf2::TransformException & ex) {
+      RCLCPP_ERROR(
+        rclcpp::get_logger("logger_name"),
+        "Could not transform init pose from Map to Odom.");
+    }
+
+    transform_initpose_once_ = false;
+  }
+
+  // Update the transformed pose in the local variable
+  odom_updated_pose_.header.frame_id = "odom";
+  odom_updated_pose_.header.stamp = this->now();
+
+  tf2::Quaternion base_link_orientation(
+    odom_updated_pose_.pose.pose.orientation.x,
+    odom_updated_pose_.pose.pose.orientation.y,
+    odom_updated_pose_.pose.pose.orientation.z,
+    odom_updated_pose_.pose.pose.orientation.w
+  );
+
+  tf2::Quaternion angular_change;
+  angular_change.setRPY(msg->angular.x * dt, msg->angular.y * dt, msg->angular.z * dt);
+  base_link_orientation *= angular_change;
+
+  odom_updated_pose_.pose.pose.orientation.x = base_link_orientation.x();
+  odom_updated_pose_.pose.pose.orientation.y = base_link_orientation.y();
+  odom_updated_pose_.pose.pose.orientation.z = base_link_orientation.z();
+  odom_updated_pose_.pose.pose.orientation.w = base_link_orientation.w();
+
+  // Forward integrate robot position and orientation using velocity commands
+  tf2::Matrix3x3 rotation_matrix(base_link_orientation);
+  tf2::Vector3 linear_velocity(msg->linear.x, msg->linear.y, msg->linear.z);
+  tf2::Vector3 rotated_linear_velocity = rotation_matrix * linear_velocity;
+
+  odom_updated_pose_.pose.pose.position.x += rotated_linear_velocity.x() * dt;
+  odom_updated_pose_.pose.pose.position.y += rotated_linear_velocity.y() * dt;
+  odom_updated_pose_.pose.pose.position.z += rotated_linear_velocity.z() * dt;
+
+  // Broadcast "odom" to "base_link" transform using the local variable
+  geometry_msgs::msg::TransformStamped odom_to_base_transform;
+  odom_to_base_transform.header.stamp = odom_updated_pose_.header.stamp;
+  odom_to_base_transform.header.frame_id = "odom";
+  odom_to_base_transform.child_frame_id = "base_link";
+  odom_to_base_transform.transform.translation.x = odom_updated_pose_.pose.pose.position.x;
+  odom_to_base_transform.transform.translation.y = odom_updated_pose_.pose.pose.position.y;
+  odom_to_base_transform.transform.translation.z = odom_updated_pose_.pose.pose.position.z;
+  odom_to_base_transform.transform.rotation = odom_updated_pose_.pose.pose.orientation;
+
+  tf_broadcaster_->sendTransform(odom_to_base_transform);
+}
+
+void LoopbackSimulator::initposeCallback(
+  const
+  geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
+{
+  RCLCPP_INFO(this->get_logger(), "Received initial pose wrt map");
+  init_pose_ = *msg;
+  init_pose_set_ = true;
+
+  if (init_odom_base_published_ == false) {
```

Comment:
Couldn't this just be in the constructor and just remove the need for this variable?

Or I suppose it could go into the timer callback if `init_pose_set_ == false` and also remove this variable and continuously publish the 0-transform until twists start coming in.

How does this handle gaps? E.g. if you have a robot navigating, then stop for a few seconds from success, then start again, do any errors pop up from TF transforms not being valid?

Reply:
Yes you are right  `init_odom_base_published_`  is not needed.

We need to have a 0-transform between "odom" and "base_footprint" only at initialization right ? So I think its okay to publish the 0-transform only once. 
Later the transform gets updated based on how the robot moves and we dont want to overwrite that transform with our script.

> if you have a robot navigating, then stop for a few seconds from success, then start again

I'll test this scenario and see what happens

Reply:
> We need to have a 0-transform between "odom" and "base_footprint" only at initialization right 

Correct, after that it should be updated by the robot's motion